sentry_sdk/integrations/starlette.py

```diff
@@ -0,0 +1,414 @@
+
+            self._exception_handlers[key] = _sentry_patched_exception_handler
+
+    middleware_class.__init__ = _sentry_middleware_init
+
+    old_call = middleware_class.__call__
+
+    async def _sentry_exceptionmiddleware_call(self, scope, receive, send):
+        # type: (Dict[str, Any], Dict[str, Any], Callable[[], Awaitable[Dict[str, Any]]], Callable[[Dict[str, Any]], Awaitable[None]]) -> None
+        # Also add the user (that was eventually set by be Authentication middle
+        # that was called before this middleware). This is done because the authentication
+        # midlleware sets the user in the scope and then (in the same function)
+        # calls this exception middelware. In case there is no exception (or no handler
+        # for the exception type occuring) then the exception bubbles up and setting the
+        # user information into the sentry scope is done in auth middleware and the
+        # ASGI middleware will then send everything to Sentry and this is fine.
+        # But if there is an exception happening that the exception middleware here
+        # has a handler for it will send the exception directly to Sentry, so we need
+        # the user information right now.
+        # This is why we do it here.
+        _add_user_to_sentry_scope(scope)
+        await old_call(self, scope, receive, send)
+
+    middleware_class.__call__ = _sentry_exceptionmiddleware_call
+
+
+def _add_user_to_sentry_scope(scope):
+    # type: (Dict[str, Any]) -> None
+    """
+    Extracts user information from the ASGI scope and
+    adds it to Sentry's scope.
+    """
+    if "user" not in scope:
+        return
+
+    hub = Hub.current
+    if hub.get_integration(StarletteIntegration) is None:
+        return
+
+    with hub.configure_scope() as sentry_scope:
+        user_info = {}  # type: Dict[str, Any]
+        starlette_user = scope["user"]
+
+        username = getattr(starlette_user, "username", None)
+        if username:
+            user_info.setdefault("username", starlette_user.username)
+
+        user_id = getattr(starlette_user, "id", None)
+        if user_id:
+            user_info.setdefault("id", starlette_user.id)
+
+        email = getattr(starlette_user, "email", None)
+        if email:
+            user_info.setdefault("email", starlette_user.email)
+
+        sentry_scope.user = user_info
+
+
+def patch_authentication_middleware(middleware_class):
+    # type: (Any) -> None
+    """
+    Add user information to Sentry scope.
+    """
+    old_call = middleware_class.__call__
+
+    async def _sentry_authenticationmiddleware_call(self, scope, receive, send):
+        # type: (Dict[str, Any], Dict[str, Any], Callable[[], Awaitable[Dict[str, Any]]], Callable[[Dict[str, Any]], Awaitable[None]]) -> None
+        try:
+            await old_call(self, scope, receive, send)
+        except Exception as exc:
+            _add_user_to_sentry_scope(scope)
+            raise exc
```

Comment:
do we still need this since you added `_add_user_to_sentry_scope` to `patch_exception_middleware` too?

Reply:
Oh, yes. I wanted to clean this up. As ExceptionMiddleware is always present, we do not need this here. Good catch!